registry/remote/repository.go

```diff
@@ -943,65 +947,214 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+	`PASS`, even though it's actually a `FAIL`; that's a function of being
+	bling, and not performing the expensive validation here for performance
+	reasons, knowing that it will be calculated at a later point in time.
+
+	 ___________________________________________________________________________________
+	| ID | CLIENT        | SERVER         | GET                   | HEAD                |
+	|----+---------------+----------------+-----------------------+---------------------+
+	| 1  | tag           | missing        | CALCULATE,PASS        | FAIL                |
+	| 2  | tag           | presentValid   | TRUST,PASS            | TRUST,PASS          |
+	| 3  | tag           | presentInvalid | TRUST,!PASS           | TRUST,!PASS         |
+	| 4  | validDigest   | missing        | CALCULATE,VERIFY,PASS | FAIL                |
+	| 5  | validDigest   | presentValid   | TRUST,COMPARE,PASS    | TRUST,COMPARE,PASS  |
+	| 6  | validDigest   | presentInvalid | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 7  | invalidDigest | missing        | CALCULATE,VERIFY,FAIL | FAIL                |
+	| 8  | invalidDigest | presentValid   | TRUST,COMPARE,FAIL    | TRUST,COMPARE,FAIL  |
+	| 9  | invalidDigest | presentInvalid | TRUST,COMPARE,!PASS   | TRUST,COMPARE,!PASS |
+	 -----------------------------------------------------------------------------------
+	*/
+	var contentDigest digest.Digest
+
+	// 5a. Set contentDigest if no early error conditions are encountered
+	var calculatedDigest digest.Digest
+	if len(hdrDigest) == 0 {
+		if httpMethod == "HEAD" {
+			// HEAD without server `Docker-Content-Digest` header is an
+			// immediate fail; HEAD
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"HTTP %s request missing required header `%s`",
+				httpMethod, dockerContentDigestHeader,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
 
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+		// GET without server `Docker-Content-Digest` header forces the
+		// expensive calculation
+		if calculatedDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		contentDigest = calculatedDigest
+	} else {
+		contentDigest = hdrDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	// 5b. With all 3 digests appropriately set, perform final 3-way validation
+	if err = verifyContentDigests(calculatedDigest, hdrDigest, refDigest); err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
 	}
```

Comment:
Do we need 3-way validation?

`calculatedDigest` has value if and only if `hdrDigest` is empty. Therefore, based on line 1047 and 1049, we only need to verify `contentDigest` and `refDigest`, which can be done easily as follows.

```go
if refDigest != "" && refDigest != contentDigest {
        return ocispec.Descriptor{}, fmt.Errorf("%s: mismatch digest: %s", refDigest, contentDigest)
}
```